File: extract_tables.js
#!/usr/bin/env node
/**
 * PDF表格提取工具 (JavaScript版本)
 * 使用pdf-parse提取PDF内容并尝试识别表格
 */

const fs = require('fs');
const path = require('path');

// 检查依赖
let pdfParse;
try {
    pdfParse = require('pdf-parse');
} catch (e) {
    console.error('❌ 错误: 缺少pdf-parse依赖');
    console.log('请运行: npm install pdf-parse');
    process.exit(1);
}

class PDFTableExtractor {
    constructor(pdfPath, outputDir = '.') {
        this.pdfPath = pdfPath;
        this.outputDir = outputDir;
        this.allTables = [];
        this.extractionInfo = {
            totalPages: 0,
            totalTables: 0,
            extractionTime: null
        };
    }

    async extractAllTables() {
        console.log(`📄 正在处理文件: ${this.pdfPath}`);
        console.log('='.repeat(60));

        if (!fs.existsSync(this.pdfPath)) {
            console.error(`❌ 错误: 文件不存在 - ${this.pdfPath}`);
            return false;
        }

        try {
            const dataBuffer = fs.readFileSync(this.pdfPath);
            
            // 使用pdf-parse直接解析
            const data = await pdfParse(dataBuffer);

            this.extractionInfo.totalPages = data.numpages;
            this.extractionInfo.extractionTime = new Date().toLocaleString('zh-CN');

            console.log(`📊 总页数: ${data.numpages}`);
            console.log(`📝 文本长度: ${data.text.length} 字符\n`);

            // 打印部分文本内容以便调试
            console.log('文本预览:');
            console.log('-'.repeat(60));
            const preview = data.text.substring(0, 500);
            console.log(preview);
            console.log('-'.repeat(60));
            console.log('');

            // 解析文本，尝试识别表格
            this.parseTables(data.text);

            this.extractionInfo.totalTables = this.allTables.length;

            return true;

        } catch (error) {
            console.error(`❌ 处理PDF时出错: ${error.message}`);
            console.error(error.stack);
            return false;
        }
    }

    parseTables(text) {
        console.log('开始解析表格...');

        const lines = text.split('\n');
        let currentTable = null;
        let tableStartPage = 1;

        for (let i = 0; i < lines.length; i++) {
            const line = lines[i].trim();

            if (!line) continue;

            // 检测表格特征：多个连续的制表符、竖线或数字密集
            const hasTabDelimiters = line.includes('\t');
            const hasPipes = line.includes('|') && line.split('|').length >= 4;
            const isDenseNumbers = /(\d+[\s\t]+){3,}/.test(line);

            // 更宽松的表格检测：如果一行有多个连续的空格分隔的词
            const multipleSpaces = line.split(/\s{2,}/).length >= 3;

            if (hasTabDelimiters || hasPipes || isDenseNumbers || multipleSpaces) {
                // 可能是表格行
                if (!currentTable) {
                    currentTable = [];
                    tableStartPage = this.estimatePage(i, lines.length, this.extractionInfo.totalPages);
                }

                // 分割列
                let columns;
                if (hasTabDelimiters) {
                    columns = line.split('\t').map(c => c.trim()).filter(c => c);
                } else if (hasPipes) {
                    columns = line.split('|').map(c => c.trim()).filter(c => c);
                } else {
                    // 按多个连续空格分割
                    columns = line.split(/\s{2,}/).map(c => c.trim()).filter(c => c);
                }

                if (columns.length > 1) {
                    currentTable.push(columns);
                }
            } else {
                // 表格结束
                if (currentTable && currentTable.length > 1) {
                    this.allTables.push({
                        page: tableStartPage,
                        data: currentTable,
                        shape: [currentTable.length, currentTable[0]?.length || 0],
                        tableId: `Table_${this.allTables.length + 1}`
                    });
                    console.log(`✅ 识别到表格 ${this.allTables.length}: ${currentTable.length}行 × ${currentTable[0]?.length || 0}列`);
                }
                currentTable = null;
            }
        }

        // 处理最后一个表格
        if (currentTable && currentTable.length > 1) {
            this.allTables.push({
                page: tableStartPage,
                data: currentTable,
                shape: [currentTable.length, currentTable[0]?.length || 0],
                tableId: `Table_${this.allTables.length + 1}`
            });
        }
    }

    estimatePage(currentLine, totalLines, totalPages) {
        // 估算当前行所在的页码
        const linesPerPage = Math.ceil(totalLines / totalPages);
        return Math.min(Math.ceil(currentLine / linesPerPage), totalPages);
    }

    exportToCSV(outputFilename = null) {
        if (this.allTables.length === 0) {
            console.log('⚠️ 没有表格数据可导出');
            return null;
        }

        if (!outputFilename) {
            const baseName = path.basename(this.pdfPath, path.extname(this.pdfPath));
            outputFilename = `${baseName}_tables.csv`;
        }

        const outputPath = path.join(this.outputDir, outputFilename);

        try {
            let csvContent = '表格来源,页码,行号,内容\n';

            for (const table of this.allTables) {
                for (let rowIdx = 0; rowIdx < table.data.length; rowIdx++) {
                    const row = table.data[rowIdx];
                    // 转义CSV中的引号和逗号
                    const escapedRow = row.map(cell =>
                        `"${String(cell).replace(/"/g, '""')}"`
                    ).join(',');
                    csvContent += `${table.tableId},${table.page},${rowIdx + 1},${escapedRow}\n`;
                }
            }

            fs.writeFileSync(outputPath, csvContent, 'utf-8');
            console.log(`\n✅ CSV文件已导出: ${outputPath}`);
            return outputPath;

        } catch (error) {
            console.error(`❌ 导出CSV时出错: ${error.message}`);
            return null;
        }
    }

    exportToJSON(outputFilename = null) {
        if (this.allTables.length === 0) {
            console.log('⚠️ 没有表格数据可导出');
            return null;
        }

        if (!outputFilename) {
            const baseName = path.basename(this.pdfPath, path.extname(this.pdfPath));
            outputFilename = `${baseName}_tables.json`;
        }

        const outputPath = path.join(this.outputDir, outputFilename);

        try {
            const exportData = {
                source: this.pdfPath,
                extractionTime: this.extractionInfo.extractionTime,
                totalPages: this.extractionInfo.totalPages,
                totalTables: this.extractionInfo.totalTables,
                tables: this.allTables
            };

            fs.writeFileSync(outputPath, JSON.stringify(exportData, null, 2), 'utf-8');
            console.log(`✅ JSON文件已导出: ${outputPath}`);
            return outputPath;

        } catch (error) {
            console.error(`❌ 导出JSON时出错: ${error.message}`);
            return null;
        }
    }

    generateReport(reportFilename = null) {
        if (!reportFilename) {
            const baseName = path.basename(this.pdfPath, path.extname(this.pdfPath));
            reportFilename = `${baseName}_report.txt`;
        }

        const reportPath = path.join(this.outputDir, reportFilename);

        try {
            let reportContent = '='.repeat(60) + '\n';
            reportContent += 'PDF 表格提取报告\n';
            reportContent += '='.repeat(60) + '\n\n';

            reportContent += `源文件: ${this.pdfPath}\n`;
            reportContent += `提取时间: ${this.extractionInfo.extractionTime}\n`;
            reportContent += `总页数: ${this.extractionInfo.totalPages}\n`;
            reportContent += `总表格数: ${this.extractionInfo.totalTables}\n\n`;

            reportContent += '-'.repeat(60) + '\n';
            reportContent += '表格详情:\n';
            reportContent += '-'.repeat(60) + '\n\n';

            for (let idx = 0; idx < this.allTables.length; idx++) {
                const table = this.allTables[idx];
                reportContent += `表格 ${idx + 1}: ${table.tableId}\n`;
                reportContent += `  位置: 第 ${table.page} 页\n`;
                reportContent += `  尺寸: ${table.shape[0]} 行 × ${table.shape[1]} 列\n`;
                
                // 添加前几行作为预览
                reportContent += `  预览:\n`;
                const previewRows = table.data.slice(0, 3);
                for (let j = 0; j < previewRows.length; j++) {
                    reportContent += `    行${j + 1}: ${JSON.stringify(previewRows[j]).substring(0, 100)}...\n`;
                }
                reportContent += '\n';
            }

            reportContent += '='.repeat(60) + '\n';

            fs.writeFileSync(reportPath, reportContent, 'utf-8');
            console.log(`📋 提取报告已生成: ${reportPath}`);
            return reportPath;

        } catch (error) {
            console.error(`⚠️ 生成报告时出错: ${error.message}`);
            return null;
        }
    }

    printSummary() {
        console.log('\n' + '='.repeat(60));
        console.log('📊 提取摘要');
        console.log('='.repeat(60));
        console.log(`源文件: ${this.pdfPath}`);
        console.log(`提取时间: ${this.extractionInfo.extractionTime}`);
        console.log(`总页数: ${this.extractionInfo.totalPages}`);
        console.log(`总表格数: ${this.extractionInfo.totalTables}`);
        console.log('='.repeat(60));
    }
}

// 主函数
async function main() {
    const pdfPath = path.join(__dirname, 'chinese_document.pdf');
    const outputDir = __dirname;

    if (!fs.existsSync(pdfPath)) {
        console.error(`❌ 错误: PDF文件不存在 - ${pdfPath}`);
        return;
    }

    // 创建提取器实例
    const extractor = new PDFTableExtractor(pdfPath, outputDir);

    // 执行提取
    if (await extractor.extractAllTables()) {
        // 打印摘要
        extractor.printSummary();

        // 导出结果
        const csvPath = extractor.exportToCSV();
        const jsonPath = extractor.exportToJSON();
        const reportPath = extractor.generateReport();

        console.log('\n' + '🎉 提取完成!' + '\n');

        if (csvPath) console.log(`📁 CSV文件: ${csvPath}`);
        if (jsonPath) console.log(`📁 JSON文件: ${jsonPath}`);
        if (reportPath) console.log(`📁 报告文件: ${reportPath}`);
    } else {
        console.log('\n❌ 提取失败，请检查错误信息');
    }
}

// 运行
main().catch(error => {
    console.error(' Fatal error:', error);
    process.exit(1);
});

File: extract_tables.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
PDF表格提取工具
从PDF文档中提取表格数据，支持中文，并导出为Excel和CSV格式
"""

import pdfplumber
import pandas as pd
import json
from datetime import datetime
import os
import sys


class PDFTableExtractor:
    """PDF表格提取器"""
    
    def __init__(self, pdf_path, output_dir="."):
        """
        初始化提取器
        
        Args:
            pdf_path: PDF文件路径
            output_dir: 输出目录
        """
        self.pdf_path = pdf_path
        self.output_dir = output_dir
        self.all_tables = []  # 存储所有表格数据
        self.extraction_info = {
            'total_pages': 0,
            'total_tables': 0,
            'tables_by_page': {},
            'extraction_time': None
        }
    
    def extract_all_tables(self):
        """提取PDF中的所有表格"""
        print(f"📄 正在处理文件: {self.pdf_path}")
        print("=" * 60)
        
        if not os.path.exists(self.pdf_path):
            print(f"❌ 错误: 文件不存在 - {self.pdf_path}")
            return False
        
        try:
            with pdfplumber.open(self.pdf_path) as pdf:
                self.extraction_info['total_pages'] = len(pdf)
                self.extraction_info['extraction_time'] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                
                print(f"📊 总页数: {len(pdf)}\n")
                
                for page_num, page in enumerate(pdf.pages, 1):
                    self._extract_tables_from_page(page, page_num)
                
                self.extraction_info['total_tables'] = len(self.all_tables)
                
                return True
                
        except Exception as e:
            print(f"❌ 处理PDF时出错: {str(e)}")
            import traceback
            traceback.print_exc()
            return False
    
    def _extract_tables_from_page(self, page, page_num):
        """从单页提取表格"""
        try:
            # 使用更精确的表格提取设置
            table_settings = {
                "vertical_strategy": "lines",
                "horizontal_strategy": "lines",
                "snap_tolerance": 5,
                "intersection_tolerance": 15,
            }
            
            # 尝试提取表格
            tables = page.extract_tables(table_settings)
            
            if not tables:
                # 如果使用lines策略没有找到表格，尝试默认策略
                tables = page.extract_tables()
            
            if tables:
                print(f"✅ 第 {page_num} 页: 发现 {len(tables)} 个表格")
                
                for table_num, table in enumerate(tables, 1):
                    # 将表格数据转换为DataFrame
                    df = pd.DataFrame(table)
                    
                    # 清洗数据：处理空值
                    df = df.fillna('').astype(str)
                    
                    # 移除完全为空的行和列
                    df = df.dropna(how='all')
                    df = df.loc[:, (df != '').any(axis=0)]
                    
                    table_info = {
                        'page': page_num,
                        'table_num': table_num,
                        'data': df,
                        'shape': df.shape,
                        'table_id': f"Page{page_num}_Table{table_num}"
                    }
                    
                    self.all_tables.append(table_info)
                    print(f"   └─ 表格 {table_num}: {df.shape[0]}行 × {df.shape[1]}列")
                
                self.extraction_info['tables_by_page'][page_num] = len(tables)
            else:
                # 即使没有表格，也记录
                print(f"⚪ 第 {page_num} 页: 未检测到表格")
                
        except Exception as e:
            print(f"⚠️ 第 {page_num} 页处理时出错: {str(e)}")
    
    def export_to_excel(self, output_filename=None):
        """导出表格到Excel文件"""
        if not self.all_tables:
            print("⚠️ 没有表格数据可导出")
            return None
        
        if output_filename is None:
            base_name = os.path.splitext(os.path.basename(self.pdf_path))[0]
            output_filename = f"{base_name}_tables.xlsx"
        
        output_path = os.path.join(self.output_dir, output_filename)
        
        try:
            with pd.ExcelWriter(output_path, engine='openpyxl') as writer:
                for table_info in self.all_tables:
                    sheet_name = table_info['table_id'][:31]  # Excel sheet名称最多31字符
                    table_info['data'].to_excel(
                        writer,
                        sheet_name=sheet_name,
                        index=False,
                        header=False  # 不使用默认的DataFrame行号
                    )
            
            print(f"\n✅ Excel文件已导出: {output_path}")
            return output_path
            
        except Exception as e:
            print(f"❌ 导出Excel时出错: {str(e)}")
            return None
    
    def export_to_csv(self, output_filename=None):
        """导出表格到CSV文件"""
        if not self.all_tables:
            print("⚠️ 没有表格数据可导出")
            return None
        
        if output_filename is None:
            base_name = os.path.splitext(os.path.basename(self.pdf_path))[0]
            output_filename = f"{base_name}_tables.csv"
        
        output_path = os.path.join(self.output_dir, output_filename)
        
        try:
            # 将所有表格合并导出到一个CSV文件
            all_data = []
            for table_info in self.all_tables:
                # 添加标识列
                df = table_info['data'].copy()
                df.insert(0, '表格来源', table_info['table_id'])
                df.insert(1, '页码', table_info['page'])
                all_data.append(df)
            
            merged_df = pd.concat(all_data, ignore_index=True)
            merged_df.to_csv(output_path, index=False, encoding='utf-8-sig')
            
            print(f"✅ CSV文件已导出: {output_path}")
            return output_path
            
        except Exception as e:
            print(f"❌ 导出CSV时出错: {str(e)}")
            return None
    
    def generate_report(self, report_filename=None):
        """生成提取报告"""
        if report_filename is None:
            base_name = os.path.splitext(os.path.basename(self.pdf_path))[0]
            report_filename = f"{base_name}_report.txt"
        
        report_path = os.path.join(self.output_dir, report_filename)
        
        try:
            with open(report_path, 'w', encoding='utf-8') as f:
                f.write("=" * 60 + "\n")
                f.write("PDF 表格提取报告\n")
                f.write("=" * 60 + "\n\n")
                
                f.write(f"源文件: {self.pdf_path}\n")
                f.write(f"提取时间: {self.extraction_info['extraction_time']}\n")
                f.write(f"总页数: {self.extraction_info['total_pages']}\n")
                f.write(f"总表格数: {self.extraction_info['total_tables']}\n\n")
                
                f.write("-" * 60 + "\n")
                f.write("表格详情:\n")
                f.write("-" * 60 + "\n\n")
                
                for idx, table_info in enumerate(self.all_tables, 1):
                    f.write(f"表格 {idx}: {table_info['table_id']}\n")
                    f.write(f"  位置: 第 {table_info['page']} 页\n")
                    f.write(f"  尺寸: {table_info['shape'][0]} 行 × {table_info['shape'][1]} 列\n\n")
                
                f.write("=" * 60 + "\n")
            
            print(f"📋 提取报告已生成: {report_path}")
            return report_path
            
        except Exception as e:
            print(f"⚠️ 生成报告时出错: {str(e)}")
            return None
    
    def print_summary(self):
        """打印提取摘要"""
        print("\n" + "=" * 60)
        print("📊 提取摘要")
        print("=" * 60)
        print(f"源文件: {self.pdf_path}")
        print(f"提取时间: {self.extraction_info['extraction_time']}")
        print(f"总页数: {self.extraction_info['total_pages']}")
        print(f"总表格数: {self.extraction_info['total_tables']}")
        print("=" * 60)


def main():
    """主函数"""
    # PDF文件路径
    pdf_path = r"d:\skill gallery\chinese_document.pdf"
    output_dir = r"d:\skill gallery"
    
    # 检查文件是否存在
    if not os.path.exists(pdf_path):
        print(f"❌ 错误: PDF文件不存在 - {pdf_path}")
        return
    
    # 创建提取器实例
    extractor = PDFTableExtractor(pdf_path, output_dir)
    
    # 执行提取
    if extractor.extract_all_tables():
        # 打印摘要
        extractor.print_summary()
        
        # 导出结果
        excel_path = extractor.export_to_excel()
        csv_path = extractor.export_to_csv()
        report_path = extractor.generate_report()
        
        print("\n" + "🎉 提取完成!" + "\n")
        
        if excel_path:
            print(f"📁 Excel文件: {excel_path}")
        if csv_path:
            print(f"📁 CSV文件: {csv_path}")
        if report_path:
            print(f"📁 报告文件: {report_path}")
    else:
        print("\n❌ 提取失败，请检查错误信息")


if __name__ == "__main__":
    main()
